README.md
- [SolisCloud to PVOutput and/or Domoticz](#soliscloud-to-pvoutput-andor-domoticz)
  - [SolisCloud](#soliscloud)
  - [PVOutput](#pvoutput)
  - [Domoticz](#domoticz)
- [Configuration](#configuration)
- [Usage: Windows 10](#usage-windows-10)
- [Usage: Linux or Raspberry pi](#usage-linux-or-raspberry-pi)
  - [Linux or Raspberry pi configuration](#linux-or-raspberry-pi-configuration)
  - [log files](#log-files)
  - [Configuration with multiple inverters in one SolisCloud station](#configuration-with-multiple-inverters-in-one-soliscloud-station)
  - [Combined data of two PVOutput accounts/inverters](#combined-data-of-two-pvoutput-accountsinverters)
- [Example standard output of SolisCloud2PVoutput](#example-standard-output-of-soliscloud2pvoutput)

# SolisCloud to PVOutput and/or Domoticz
Simple Python3 script to copy latest (normally once per 5 minutes) SolisCloud portal inverter update to PVOutput portal and/or Domoticz.

The soliscloud_to_pvoutput.py script will get the first station id with the secrets of SolisCloud (see next section). Thereafter it will get the inverter id and serial number via the configured SOLISCLOUD_INVERTER_INDEX (default the first inverter). Then in an endless loop the inverter details are fetched and the following information is used:
* timestamp
* DC PV voltage (assuming no more than 4 strings)
* watt (current)
* watthour today
* inverter temperature (instead of outside temperature, you can still overrule with weather device)
* AC voltage (max voltage of 3 phases), default "Power Consumption" field is misused, so read "AC Volt" for the "Power Used" column of PVOutput and ignore "Energy Used" column
* People with [Solis Consumption Monitoring solution installed](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/issues/15) can also send the Power Used (FamilyLoadPower) instead of AC Voltage
* People with [Solis Consumption Monitoring solution and Solis hybrid inverter with battery storage](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/discussions/24) can also send the Power Used (HomeConsumption) instead of AC Voltage.

This information is used to compute the new information to be send to PVOutput and/or Domoticz, when the timestamp is changed.

Notes
* only between 5 and 23 hour data is fetched from SolisCloud and copied to PVOutput and/or Domoticz
* the script will exit outside 5 and 23
* Each new day the "watthour today" starts with 0
* Because the resolution of the SolisCloud watthour is in 100 Watt, a higher resolution is computed with current Watt
* if you have more than 1 station [you need to configure each inverter in different directories](#configuration-with-multiple-inverters-in-one-soliscloud-station)
* If you have more than 4 strings or a 3 phase inverter, you need to adapt the script (sorry, not supported yet)

## SolisCloud
[SolisCloud](https://www.soliscloud.com/) is the next generation Portal for Solis branded PV systems from Ginlong.

The python script requires a SolisCloud API_ID, API_SECRET and API_URL to function.
* Go to https://www.soliscloud.com/#/apiManage
* Ativate API management and agree with the usage conditions.
* After activation, click on view key tot get a pop-up window asking for the verification code.
* First click on "Verification code" after which you get an image with 2 puzzle pieces, which you need to overlap each other using the slider below.
* After that, you will receive an email with the verification code you need to enter (within 60 seconds).
* Once confirmed, you get the API_ID, API_SECRET and API_URL

## PVOutput
[PVOutput](https://pvoutput.org/) is a free online service for sharing and comparing photovoltaic solar panel output data. It provides both manual and automatic data uploading facilities.

Output data can be graphed, analysed and compared with other pvoutput contributors over various time periods. The ability to compare with similar systems within close proximity allows both short and longer term performance issues to be easily identified. While PVOutput is primarily focused on monitoring energy generation, it also provides equally capabable facilities to upload and monitor energy consumption data from various energy monitoring devices.

The python script requires a PVOutput API_KEY and SYSTEM_ID to function.
* Login in PVOutput and goto your Settings page
* Select Enabled for API Access
* Click on New Key to generate your API key
* Make a note of your System Id
* Save your settings

## Domoticz
[Domoticz](https://www.domoticz.com/) is a very light weight home automation system that lets you monitor and configure miscellaneous devices, including lights, switches, various sensors/meters like temperature, rainfall, wind, ultraviolet (UV) radiation, electricity usage/production, gas consumption, water consumption and many more. Notifications/alerts can be sent to any mobile device.

If you want to know how to configure in Domoticz your inverter, see [this discussion](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/discussions/18).

![alt text](https://user-images.githubusercontent.com/17342657/237064859-b7bcb83a-a753-4399-b60d-801bdd2741a3.png)

![alt text](https://user-images.githubusercontent.com/17342657/237064582-59fcd74b-5b04-4578-98a4-18819bf8482f.png)

# Configuration
Change in soliscloud_to_pvoutput.cfg the following lines with your above obtained secrets and domoticz configuration, including if you want to send to PVOutput, Domoticz or both. By default only output is send to PVOutput:
````
[api_secrets]
soliscloud_api_id = 1300386381123456789
soliscloud_api_secret = 304abf2bd8a44242913d704123456789
soliscloud_api_url = https://www.soliscloud.com:13333
soliscloud_inverter_index = 0
pvoutput_api_key = 0f2dd8190d00369ec893b059034dde1123456789
pvoutput_system_id = 12345

[PVOutput]
send_to_pvoutput = True
pvoutput_fill_temperature_with_inverter_temperature = True
pvoutput_fill_voltage_with_ac_voltage = False
pvoutput_fill_power_consumption_with_familyloadpower = False
pvoutput_fill_power_consumption_with_homeconsumption = False
pvoutput_fill_power_consumption_with_ac_voltage = True

[Domoticz]
send_to_domoticz = False
domot_url = http://192.168.0.222:8081
domot_power_generated_id = 0
domot_ac_volt_id = 0
domot_inverter_temp_id = 0
domot_volt_id = 0
domot_solarpower_id = 0
domot_energygeneration_id = 0
domot_batterypower_id = 0
domot_gridpower_id = 0
domot_familyloadpower_id = 0
domot_homeconsumption_id = 0
````

Because I see some forks or local adaptions for people wanting a slightly different behavior, I made some adaptions to the SolisCloud2PVOutput solution and configuration to capture (some of) those variations.
Now the following is possible:
- made it possible to have the config files not in the current directory, but just use the directory where the python script is located
- use AC voltage instead of DC Voltage for the PVOutput Voltage field (setting pvoutput_fill_voltage_with_ac_voltage)
- do NOT send the inverter temperature to PVOutput (setting pvoutput_fill_temperature_with_inverter_temperature)
- extra values possible to send to Domoticz (see above), when domot_[name]_id is 0, it will NOT be send to domoticz
- instead of misusing the PVOutput PowerConsumption field for the AC voltage, you can also use FamilyLoadPower or HomeConsumption (setting pvoutput_fill_power_consumption_with_familyloadpower, pvoutput_fill_power_consumption_with_homeconsumption, pvoutput_fill_power_consumption_with_ac_voltage)

Note 1: for the last bullet, you need to have a [Solis Consumption Monitoring solution installed](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/issues/15), also supported is [a Solis hybrid inverter with battery storage](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/discussions/24).

Note 2: make sure that you move send_to_pvoutput setting to the [PVOutput] section, if you have an already existing configuration.



# Usage: Windows 10
Make sure to go to the directory where soliscloud_to_pvoutput.py and soliscloud_to_pvoutput.cfg is located.
````
python soliscloud_to_pvoutput.py
````

# Usage: Linux or Raspberry pi
soliscloud_to_pvoutput.py scripts runs on my Raspberry pi with Raspbian GNU/Linux 11 (bullseye).

## Linux or Raspberry pi configuration
Steps:
* create a directory solis in your home directory
* copy solis.sh, soliscloud_to_pvoutput.py, soliscloud_to_pvoutput.cfg and logging_config.ini in this solis directory
* change inside soliscloud_to_pvoutput.cfg the API secrets
* chmod +x solis.sh
* add the following line in your crontab -e:

```
2 5 * * * ~/solis/solis.sh > /dev/null
@reboot sleep 123 && ~/solis/solis.sh > /dev/null
```

## log files
Log files are written in the home subdirectory solis
* solis.log containing the data send to PVOutput (and maybe error messages)
* solis.crontab.log containing the crontab output (normally it will say that solis.sh is running).

## Configuration with multiple inverters in one SolisCloud station

Make 2 PVOutput accounts (you need 2 email addresses) for each inverter a separate PVOutput account. Make sure to configure the PVOutput accounts and get the PVOutput API keys.

The solution is to have 2 scripts running in different directories (one for each inverter) and for the each directory you do modifications, e.g. the configuration to get the appropriate inverter (setting soliscloud_inverter_index) and send the output to a appropriate PVOutput account as target.

Create two directories, copy the SolisCloud2PVOutput files (soliscloud_to_pvoutput.py, soliscloud_to_pvoutput.cfg, solis.sh and logging_config.ini) to each directory and configure in each directory soliscloud_to_pvoutput.cfg:
- solis
- solis2

In solis2 directory you change the following:
- modify soliscloud_to_pvoutput.cfg to point the second PVOutput account secrets and change the soliscloud_inverter_index to 1 (to get the data of the second inverter)
- rename solis.sh to solis2.sh and modify solis2.sh to go to directory solis2 (line 9: cd ~/solis2)

Have two cronrabs running (for solis.sh and solis2.sh)

## Combined data of two PVOutput accounts/inverters

if you also want the combined data of the two inverters, use a third PVOutput account (yet another email address) and use my python tool [CombinePVOutputSystems](https://github.com/ZuinigeRijder/CombinePVOutputSystems#combine-pvoutput-systems).

# Example standard output of SolisCloud2PVoutput

```
20220730 23:00:17: Outside solar generation hours (5..23)
Exiting program to start fresh tomorrow
....
20220904 17:52:32: data=20220904,17:50,7600,90,-1,227.7,36.5,161.5
20220904 17:56:34: data=20220904,17:55,7600,100,-1,227.3,36.4,161.5
20220904 18:01:42: data=20220904,18:00,7600,100,-1,226.7,36.2,161.4
20220904 18:05:46: data=20220904,18:05,7700,100,-1,226.9,36.1,161.4
20220904 18:10:51: data=20220904,18:10,7700,110,-1,227.4,36.1,161.4

```


CHANGELOG.md
<a name="R1.13.0"></a>
# [Adaptions to the SolisCloud2PVOutput solution and configuration to support variations in use (R1.13.0)](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/releases/tag/R1.13.0) - 22 Jun 2023

Because I see some forks or local adaptions for people wanting a slightly different behavior, I made some adaptions to the SolisCloud2PVOutput solution and configuration to capture (some of) those variations.
Now the following is possible:
- made it possible to have the config files not in the current directory, but just use the directory where the python script is located
- use AC voltage instead of DC Voltage for the PVOutput Voltage field (setting pvoutput_fill_voltage_with_ac_voltage)
- do NOT send the inverter temperature to PVOutput (setting pvoutput_fill_temperature_with_inverter_temperature)
- extra values possible to send to Domoticz (see above), when domot_[name]_id is 0, it will NOT be send to domoticz
- instead of misusing the PVOutput PowerConsumption field for the AC voltage, you can also use FamilyLoadPower or HomeConsumption (setting pvoutput_fill_power_consumption_with_familyloadpower, pvoutput_fill_power_consumption_with_homeconsumption, pvoutput_fill_power_consumption_with_ac_voltage)

Note 1: for the last bullet, you need to have a [Solis Consumption Monitoring solution installed](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/issues/15), also supported is [a Solis hybrid inverter with battery storage](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/discussions/24).

Note 2: make sure that you move send_to_pvoutput setting to the [PVOutput] section, if you have an already existing configuration.

Change in soliscloud_to_pvoutput.cfg the following lines with your above obtained secrets and domoticz configuration, including if you want to send to PVOutput, Domoticz or both. By default only output is send to PVOutput:
````
[api_secrets]
soliscloud_api_id = 1300386381123456789
soliscloud_api_secret = 304abf2bd8a44242913d704123456789
soliscloud_api_url = https://www.soliscloud.com:13333
soliscloud_inverter_index = 0
pvoutput_api_key = 0f2dd8190d00369ec893b059034dde1123456789
pvoutput_system_id = 12345

[PVOutput]
send_to_pvoutput = True
pvoutput_fill_temperature_with_inverter_temperature = True
pvoutput_fill_voltage_with_ac_voltage = False
pvoutput_fill_power_consumption_with_familyloadpower = False
pvoutput_fill_power_consumption_with_homeconsumption = False
pvoutput_fill_power_consumption_with_ac_voltage = True

[Domoticz]
send_to_domoticz = False
domot_url = http://192.168.0.222:8081
domot_power_generated_id = 0
domot_ac_volt_id = 0
domot_inverter_temp_id = 0
domot_volt_id = 0
domot_solarpower_id = 0
domot_energygeneration_id = 0
domot_batterypower_id = 0
domot_gridpower_id = 0
domot_familyloadpower_id = 0
domot_homeconsumption_id = 0
````


[Changes][R1.13.0]


<a name="R1.12.0"></a>
# [Avoid exiting program when SolisCloud returns not filled data (R1.12.0)](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/releases/tag/R1.12.0) - 31 May 2023

I encounter that SolisCloud returned data without error, but the data was not filled correctly. As  a result an exception occurred and the program exited. Fortunately I noticed it, so I could start the script the same day again.

Added a try/catch and when such an exception occurs, wait one minute and try again.

Also renamed some variables for better readability.

[Changes][R1.12.0]


<a name="R1.11.0"></a>
# [Added support for multiple inverters (R1.11.0)](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/releases/tag/R1.11.0) - 12 May 2023

See this [discussion](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/discussions/23).

Added configuration item (make sure to add this new configuration item to soliscloud_to_pvoutput.cfg if you have an existing configuration) :
- soliscloud_inverter_index = 0

Added below information about multiple inverters to the README.md.


## Domoticz
[Domoticz](https://www.domoticz.com/) is a very light weight home automation system that lets you monitor and configure miscellaneous devices, including lights, switches, various sensors/meters like temperature, rainfall, wind, ultraviolet (UV) radiation, electricity usage/production, gas consumption, water consumption and many more. Notifications/alerts can be sent to any mobile device.

If you want to know how to configure in Domoticz your inverter, see [this discussion](https://github.com/ZuinigeRijder/SolisCloud2PVOutput/discussions/18).

![alt text](https://user-images.githubusercontent.com/17342657/237064859-b7bcb83a-a753-4399-b60d-801bdd2741a3.png)

![alt text](https://user-images.githubusercontent.com/17342657/237064582-59fcd74b-5b04-4578-98a4-18819bf8482f.png)

## Configuration with multiple inverters in one SolisCloud station

Make 2 PVOutput accounts (you need 2 email addresses) for each inverter a separate PVOutput account. Make sure to configure the PVOutput accounts and get the PVOutput API keys.

The solution is to have 2 scripts running in different directories (one for each inverter) and for the each directory you do modifications, e.g. the configuration to get the appropriate inverter and send the output to a appropriate PVOutput account as target.

Create two directories, copy the SolisCloud2PVOutput files to each directory and configure in each directory soliscloud_to_pvoutput.cfg:
- solis
- solis2

In solis2 directory you change the following:
- modify soliscloud_to_pvoutput.cfg to point the second PVOutput account secrets and change the soliscloud_inverter_index to 1 (to get the data of the second inverter)
- rename solis.sh to solis2.sh and modify solis2.sh to go to directory solis2 (line 9: cd ~/solis2)

Have two cronrabs running (for solis.sh and solis2.sh)

## Combined data of two PVOutput accounts/inverters


soliscloud_to_pvoutput.py
# == soliscloud_to_pvoutput.py Author: Zuinige Rijder =========================
"""
Simple Python3 script to copy latest
(normally once per 5 minutes) SolisCloud portal update to PVOutput portal.
 """
from os import path
import base64
import hashlib
import hmac
import json
import time
import sys
import configparser
import socket
import traceback
import logging
import logging.config

from datetime import datetime, timezone
from urllib.error import HTTPError, URLError
from urllib.request import urlopen, Request

# == read api_secrets in soliscloud_to_pvoutput.cfg ==========================
SCRIPT_DIRNAME = path.abspath(path.dirname(__file__))
parser = configparser.ConfigParser()
parser.read(f"{SCRIPT_DIRNAME}/soliscloud_to_pvoutput.cfg")


def get(dictionary: dict, key: str, default: str = "") -> str:
    """get key from dictionary"""
    if key in dictionary:
        return dictionary[key].strip()
    return default


def get_bool(dictionary: dict, key: str, default: bool = True) -> bool:
    """get boolean value from dictionary key"""
    value = get(dictionary, key)
    if value == "":
        return default
    return value.lower() == "true"


# == API Secrets, fill in yours in soliscloud_to_pvoutput.cfg ================
api_secrets = dict(parser.items("api_secrets"))
SOLISCLOUD_API_ID = get(api_secrets, "soliscloud_api_id")  # userId
SOLISCLOUD_API_SECRET = get(api_secrets, "soliscloud_api_secret").encode("utf-8")
SOLISCLOUD_API_URL = get(api_secrets, "soliscloud_api_url")
SOLISCLOUD_INVERTER_INDEX = int(get(api_secrets, "soliscloud_inverter_index", "0"))
PVOUTPUT_API_KEY = get(api_secrets, "pvoutput_api_key")
PVOUTPUT_SYSTEM_ID = get(api_secrets, "pvoutput_system_id")

# == PVOutput info, fill in yours in soliscloud_to_pvoutput.cfg ===========
pvoutput_info = dict(parser.items("PVOutput"))
SEND_TO_PVOUTPUT = get_bool(pvoutput_info, "send_to_pvoutput")  # default True
PVOUTPUT_FILL_TEMPERATURE = get_bool(
    pvoutput_info, "pvoutput_fill_temperature_with_inverter_temperature"
)
PVOUTPUT_FILL_VOLTAGE_WITH_AC_VOLTAGE = get_bool(
    pvoutput_info, "pvoutput_fill_voltage_with_ac_voltage", False
)

PVOUTPUT_FILL_POWER_CONSUMPTION_WITH_FAMILYLOADPOWER = get_bool(
    pvoutput_info, "pvoutput_fill_power_consumption_with_familyloadpower"
)
PVOUTPUT_FILL_POWER_CONSUMPTION_WITH_HOMECONSUMPTION = get_bool(
    pvoutput_info, "pvoutput_fill_power_consumption_with_homeconsumption"
)
PVOUTPUT_FILL_POWER_CONSUMPTION_WITH_AC_VOLTAGE = get_bool(
    pvoutput_info, "pvoutput_fill_power_consumption_with_ac_voltage"
)


# == domoticz info, fill in yours in soliscloud_to_pvoutput.cfg ===========
domoticz_info = dict(parser.items("Domoticz"))
SEND_TO_DOMOTICZ = get_bool(domoticz_info, "send_to_domoticz", False)
DOMOTICZ_URL = get(domoticz_info, "domot_url")
DOMOTICZ_POWER_GENERATED_ID = get(domoticz_info, "domot_power_generated_id", "0")
DOMOTICZ_AC_VOLT_ID = get(domoticz_info, "domot_ac_volt_id", "0")
DOMOTICZ_INVERTER_TEMP_ID = get(domoticz_info, "domot_inverter_temp_id", "0")
DOMOTICZ_VOLT_ID = get(domoticz_info, "domot_volt_id", "0")
DOMOTICZ_SOLARPOWER_ID = get(domoticz_info, "domot_solarpower_id", "0")
DOMOTICZ_ENERGYGENERATION_ID = get(domoticz_info, "domot_energygeneration_id", "0")
DOMOTICZ_BATTERYPOWER_ID = get(domoticz_info, "domot_batterypower_id", "0")
DOMOTICZ_GRIDPOWER_ID = get(domoticz_info, "domot_gridpower_id", "0")
DOMOTICZ_FAMILYLOADPOWER_ID = get(domoticz_info, "domot_familyloadpower_id", "0")
DOMOTICZ_HOMECONSUMPTION_ID = get(domoticz_info, "domot_homeconsumption_id", "0")

# == Constants ===============================================================
VERB = "POST"
CONTENT_TYPE = "application/json"
USER_STATION_LIST = "/v1/api/userStationList"
INVERTER_LIST = "/v1/api/inverterList"
INVERTER_DETAIL = "/v1/api/inverterDetail"
PVOUTPUT_ADD_URL = "http://pvoutput.org/service/r2/addbatchstatus.jsp"


TODAY = datetime.now().strftime("%Y%m%d")  # format yyyymmdd

logging.config.fileConfig(f"{SCRIPT_DIRNAME}/logging_config.ini")


# == post ====================================================================
def execute_request(url: str, data: str, headers: dict) -> str:
    """execute request and handle errors"""
    if data != "":
        post_data = data.encode("utf-8")
        request = Request(url, data=post_data, headers=headers)
    else:
        request = Request(url)
    errorstring = ""
    try:
        with urlopen(request, timeout=30) as response:
            body = response.read()
            content = body.decode("utf-8")
            logging.debug(content)
            return content
    except HTTPError as error:
        errorstring = str(error.status) + ": " + error.reason
    except URLError as error:
        errorstring = str(error.reason)
    except TimeoutError:
        errorstring = "Request timed out"
    except socket.timeout:
        errorstring = "Socket timed out"
    except Exception as ex:  # pylint: disable=broad-except
        errorstring = "urlopen exception: " + str(ex)
        traceback.print_exc()

    logging.error(url + " -> " + errorstring)
    time.sleep(60)  # retry after 1 minute
    return "ERROR"


# == get_solis_cloud_data ====================================================
def get_solis_cloud_data(url_part: str, data: str) -> str:
    """get solis cloud data"""
    md5 = base64.b64encode(hashlib.md5(data.encode("utf-8")).digest()).decode("utf-8")
    while True:
        now = datetime.now(timezone.utc).strftime("%a, %d %b %Y %H:%M:%S GMT")
        encrypt_str = (
            VERB + "\n" + md5 + "\n" + CONTENT_TYPE + "\n" + now + "\n" + url_part
        )
        hmac_obj = hmac.new(
            SOLISCLOUD_API_SECRET,
            msg=encrypt_str.encode("utf-8"),
            digestmod=hashlib.sha1,
        )
        authorization = (
            "API "
            + SOLISCLOUD_API_ID
            + ":"
            + base64.b64encode(hmac_obj.digest()).decode("utf-8")
        )
        headers = {
            "Content-MD5": md5,
            "Content-Type": CONTENT_TYPE,
            "Date": now,
            "Authorization": authorization,
        }
        content = execute_request(SOLISCLOUD_API_URL + url_part, data, headers)
        # log(SOLISCLOUD_API_URL+url_part + "->" + content)
        if content != "ERROR":
            return content


# == send_pvoutput_data ======================================================
def send_pvoutput_data(pvoutput_string: str):
    """send pvoutput data with the provided parameters"""
    logging.info(pvoutput_string)
    headers = {
        "X-Pvoutput-Apikey": PVOUTPUT_API_KEY,
        "X-Pvoutput-SystemId": PVOUTPUT_SYSTEM_ID,
        "Content-Type": "application/x-www-form-urlencoded",
        "Accept": "text/plain",
    }
    retry = 0
    while True:
        retry += 1
        content = execute_request(PVOUTPUT_ADD_URL, pvoutput_string, headers)
        if content != "ERROR" or retry > 30:
            if content == "ERROR":
                logging.error("number of retries exceeded")
            return


# == send to Domoticz ========================================================
def send_to_domoticz(idx: str, value: str):
    """send_to_Domoticz"""
    if idx == "0":
        return  # nothing to do

    url = (
        DOMOTICZ_URL
        + "/json.htm?type=command&param=udevice&idx="
        + idx
        + "&svalue="
        + value
    )
    logging.info(url)
    retry = 0
    while True:
        retry += 1
        content = execute_request(url, "", {})
        if content != "ERROR" or retry > 30:
            if content == "ERROR":
                logging.error("number of retries exceeded")
            return


# == get_inverter_list_body ==================================================
def get_inverter_list_body() -> str:
    """get inverter list body"""
    body = '{"userid":"' + SOLISCLOUD_API_ID + '"}'
    content = get_solis_cloud_data(USER_STATION_LIST, body)
    station_info = json.loads(content)["data"]["page"]["records"][0]
    station_id = station_info["id"]

    body = '{"stationId":"' + station_id + '"}'
    content = get_solis_cloud_data(INVERTER_LIST, body)
    inverter_info = json.loads(content)["data"]["page"]["records"][
        SOLISCLOUD_INVERTER_INDEX
    ]
    inverter_id = inverter_info["id"]
    inverter_sn = inverter_info["sn"]

    body = '{"id":"' + inverter_id + '","sn":"' + inverter_sn + '"}'
    logging.info("body: %s", body)
    return body


# == do_work ====================================================================
def do_work():
    """do_work"""
    inverter_detail_body = get_inverter_list_body()
    timestamp_previous = "0"
    energy_generation = 0
    while True:
        time.sleep(60)  # wait 1 minute before checking again
        datetime_now = datetime.now()
        # only check between 5 and 23 hours
        if datetime_now.hour < 5 or datetime_now.hour > 22:
            logging.info("Outside solar generation hours (5..23)")
            sys.exit("Exiting program to start fresh tomorrow")

        content = get_solis_cloud_data(INVERTER_DETAIL, inverter_detail_body)
        inverter_detail = json.loads(content)["data"]
        # json_formatted_str = json.dumps(inverter_detail, indent=2)
        # print(json_formatted_str)
        timestamp_current = inverter_detail["dataTimestamp"]
        dc_voltage = str(
            inverter_detail["uPv1"]
            + inverter_detail["uPv2"]
            + inverter_detail["uPv3"]
            + inverter_detail["uPv4"]
        )
        solar_watthour_today = round(inverter_detail["eToday"] * 1000)
        inverter_temperature = str(inverter_detail["inverterTemperature"])
        ac_voltage = str(
            max(
                inverter_detail["uAc1"],
                inverter_detail["uAc2"],
                inverter_detail["uAc3"],
            )
        )
        solar_power = round(inverter_detail["pac"] * 1000)
        family_load = round(inverter_detail["familyLoadPower"] * 1000)
        battery_power = round(inverter_detail["batteryPower"] * 1000)
        grid_power = round(inverter_detail["psum"] * 1000)
        home_consumption = solar_power - grid_power - battery_power
        if timestamp_previous == "0":
            energy_generation = solar_watthour_today

        # compute local time, soliscloud did not take care of leap year
        datetime_current = datetime.fromtimestamp(int(timestamp_current) / 1000)
        if (
            timestamp_current != timestamp_previous
            and datetime_current.day == datetime_now.day
        ):
            # only handle new values today
            if timestamp_previous != "0":  # check for multiple of 5 minutes
                # round to nearest 5 minutes
                elapsed_minutes = round(
                    (int(timestamp_current) - int(timestamp_previous)) / 60000
                )
                if elapsed_minutes <= 0:
                    logging.error(  # pylint:disable=logging-fstring-interpolation
                        f"TIMESTAMPERROR: {elapsed_minutes}, timestamp: {timestamp_current}, timestamp_previous: {timestamp_previous}"  # noqa
                    )
                    elapsed_minutes = 1

                # compute hiRes energy_generation with current watts/elapsed minutes
                energy_generation += int(solar_power / (60 / elapsed_minutes))
                if energy_generation < solar_watthour_today:
                    energy_generation = solar_watthour_today  # too low
                else:
                    if solar_watthour_today + 100 < energy_generation:
                        energy_generation = solar_watthour_today + 99  # too high

            current_time = datetime_current.strftime("%H:%M")
            if logging.DEBUG >= logging.root.level:
                debug_string = f"date={TODAY}, time={current_time}, energy_generation={energy_generation}, solar_power={solar_power}, battery_power={battery_power}, grid_power={grid_power}, family_load={family_load}, home_consumption={home_consumption}, inverter_temperature={inverter_temperature}, dc_voltage={dc_voltage}, ac_voltage={ac_voltage}"  # noqa
                logging.debug(debug_string)

            if SEND_TO_PVOUTPUT:
                energy_consumption = ""  # no energy consumption
                power_consumption = ""
                if PVOUTPUT_FILL_POWER_CONSUMPTION_WITH_FAMILYLOADPOWER:
                    power_consumption = str(family_load)
                elif PVOUTPUT_FILL_POWER_CONSUMPTION_WITH_HOMECONSUMPTION:
                    power_consumption = str(home_consumption)
                elif PVOUTPUT_FILL_POWER_CONSUMPTION_WITH_AC_VOLTAGE:
                    power_consumption = ac_voltage
                temperature = ""
                if PVOUTPUT_FILL_TEMPERATURE:
                    temperature = inverter_temperature  # inverter temp iso outside temp
                voltage = dc_voltage
                if PVOUTPUT_FILL_VOLTAGE_WITH_AC_VOLTAGE:
                    voltage = ac_voltage

                # PVOutput Add Batch Status Service fields
                #
                # Field                 Required	Format	    Unit	    Example
                # Date	                Yes	        yyyymmdd	date	    20210228
                # Time	                Yes	        hh:mm	    time	    13:00
                # Energy Generation	    Yes	        number	    watt hours	10000
                # Power Generation	    No	        number	    watts	    2000
                # Energy Consumption	No	        number	    watt hours	10000
                # Power Consumption	    No	        number	    watts	    2000
                # Temperature	        No	        decimal	    celsius	    23.4
                # Voltage	            No	        decimal	    volts	    240.7

                pvoutput_string = f"data={TODAY},{current_time},{energy_generation},{solar_power},{energy_consumption},{power_consumption},{temperature},{voltage}"  # noqa
                send_pvoutput_data(pvoutput_string)

            if SEND_TO_DOMOTICZ:
                send_to_domoticz(
                    DOMOTICZ_POWER_GENERATED_ID,
                    str(solar_power) + ";" + str(energy_generation),
                )
                send_to_domoticz(DOMOTICZ_AC_VOLT_ID, ac_voltage)
                send_to_domoticz(DOMOTICZ_INVERTER_TEMP_ID, inverter_temperature)
                send_to_domoticz(DOMOTICZ_VOLT_ID, dc_voltage)
                send_to_domoticz(DOMOTICZ_SOLARPOWER_ID, str(solar_power))
                send_to_domoticz(DOMOTICZ_ENERGYGENERATION_ID, str(energy_generation))
                send_to_domoticz(DOMOTICZ_BATTERYPOWER_ID, str(battery_power))
                send_to_domoticz(DOMOTICZ_GRIDPOWER_ID, str(grid_power))
                send_to_domoticz(DOMOTICZ_FAMILYLOADPOWER_ID, str(family_load))
                send_to_domoticz(DOMOTICZ_HOMECONSUMPTION_ID, str(home_consumption))
            timestamp_previous = timestamp_current


def main_loop():
    """main_loop"""
    finished = False
    while not finished:
        try:
            do_work()
            logging.info("Progam finished successful")
            finished = True
        except Exception as exception:  # pylint: disable=broad-except
            logging.error(  # pylint:disable=logging-fstring-interpolation
                f"Exception: {exception}, sleeping a minute"
            )
            traceback.print_exc()
            time.sleep(60)


# == MAIN ====================================================================
main_loop()
